Mathlib/Analysis/Fourier/Notation.lean

```diff
@@ -0,0 +1,105 @@
+/-
+Copyright (c) 2025 Moritz Doll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Moritz Doll
+-/
+import Mathlib.Algebra.Module.LinearMap.Defs
+
+/-! # Type classes for the Fourier transform
+
+In this file we define type classes for the Fourier transform and the inverse Fourier transform.
+We introduce the notation `𝓕` and `𝓕⁻` in these classes to denote the Fourier transform and
+the inverse Fourier transform, respectively.
+
+Moreover, we provide type-classes that encode the linear structure and the Fourier inversion
+theorem.
+-/
+
+universe u v w
+
+/--
+The notation typeclass for the Fourier transform.
+-/
+class FourierTransform (E : Type u) (F : outParam (Type v)) where
+  /-- `𝓕 f` is the Fourier transform of `f`. The meaning of this notation is type-dependent. -/
+  fourierTransform : E → F
+
+/--
+The notation typeclass for the inverse Fourier transform.
+-/
+class FourierTransformInv (E : Type u) (F : outParam (Type v)) where
+  /-- `𝓕⁻ f` is the inverse Fourier transform of `f`. The meaning of this notation is
+  type-dependent. -/
+  fourierTransformInv : E → F
+
+namespace FourierTransform
+
+@[inherit_doc] scoped notation "𝓕" => FourierTransform.fourierTransform
+@[inherit_doc] scoped notation "𝓕⁻" => FourierTransformInv.fourierTransformInv
+
+end FourierTransform
+
+section add
+
+open FourierTransform
+
+/-- A `FourierModule` is a function space on which the Fourier transform is a linear map. -/
+class FourierModule (R : Type*) (E : Type*) (F : outParam (Type*)) [Add E] [Add F] [SMul R E]
+    [SMul R F] extends FourierTransform E F where
+  fourier_add : ∀ (f g : E), 𝓕 (f + g) = 𝓕 f + 𝓕 g
+  fourier_smul : ∀ (r : R) (f : E), 𝓕 (r • f) = r • 𝓕 f
+
+attribute [simp] FourierModule.fourier_add
+attribute [simp] FourierModule.fourier_smul
+
+/-- A `FourierInvModule` is a function space on which the Fourier transform is a linear map. -/
+class FourierInvModule (R : Type*) (E : Type*) (F : outParam (Type*)) [Add E] [Add F] [SMul R E]
+    [SMul R F] extends FourierTransformInv E F where
+  fourierInv_add : ∀ (f g : E), 𝓕⁻ (f + g) = 𝓕⁻ f + 𝓕⁻ g
+  fourierInv_smul : ∀ (r : R) (f : E), 𝓕⁻ (r • f) = r • 𝓕⁻ f
+
+attribute [simp] FourierInvModule.fourierInv_add
+attribute [simp] FourierInvModule.fourierInv_smul
```

Comment:
In what circumstances do we actually have a fourier transform or inverse where they aren't linear? Can we just bake these into the notation typeclass? Or is your goal to achieve better performance by separating them?

Reply:
Yeah, this is the one that I realized was blocking. Thanks. Both reasons make sense now.